core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -545,9 +551,18 @@ class LogManager(logDirs: Array[File],
         if (!isNew && offlineLogDirs.nonEmpty)
           throw new KafkaStorageException(s"Can not create log for $topicPartition because log directories ${offlineLogDirs.mkString(",")} are offline")
 
-        val dataDir = nextLogDir()
+        val logDir = {
+          val preferredLogDir = preferredLogDirs.get(topicPartition)
+          if (preferredLogDir != null)
+            preferredLogDir
+          else
+            nextLogDir().getAbsolutePath
+        }
+        if (!isLogDirOnline(logDir))
+          throw new KafkaStorageException(s"Can not create log for $topicPartition because log directory $logDir is offline")
```

Comment:
After double checking the logic, I think we will actually return `KafkaStorageException` for `OffsetByTimestamp` query if log creation failed. This logic is enforced in KIP-112. More specifically, after broker receives LeaderAndIsrRequest to create the log, if `LogManager.getOrCreateLog()` throws `KafkaStorageException`, `Partition.getOrCreateReplica()` will also throw exception without putting this replica in the `Partition.assignedReplicaMap`. 

Later we will execute the following code in `ReplicaManager.becomeLeaderOrFollower()`:

```
leaderAndISRRequest.partitionStates.asScala.keys.foreach( topicPartition =>
  /*
   * If there is offline log directory, a Partition object may have been created by getOrCreatePartition()
   * before getOrCreateReplica() failed to create local replica due to KafkaStorageException.
   * In this case ReplicaManager.allPartitions will map this topic-partition to an empty Partition object.
   * we need to map this topic-partition to OfflinePartition instead.
   */
  if (getReplica(topicPartition).isEmpty && (allPartitions.get(topicPartition) ne ReplicaManager.OfflinePartition))
    allPartitions.put(topicPartition, ReplicaManager.OfflinePartition)
)
```

This logic makes sure that this partition will map to `ReplicaManager.OfflinePartition` when this broker receives any request from user, including requests for OffsetByTimestamp. The response will return `KafkaStorageException` to the user.